core/src/cell.rs

```diff
@@ -56,90 +55,81 @@ pub struct ResolvedTransaction {
-        let mut seen_inputs = HashSet::new();
-
-        let input_cells = transaction
-            .input_pts()
-            .iter()
-            .map(|input| {
-                if seen_inputs.insert(input.clone()) {
-                    self.cell(input)
-                } else {
-                    CellStatus::Dead
-                }
-            })
-            .collect();
-
-        let dep_cells = transaction
-            .dep_pts()
-            .iter()
-            .map(|dep| {
-                if seen_inputs.insert(dep.clone()) {
-                    self.cell(dep)
-                } else {
-                    CellStatus::Dead
-                }
-            })
-            .collect();
-
-        ResolvedTransaction {
-            transaction: transaction.clone(),
-            input_cells,
-            dep_cells,
-        }
+        let mut seen_inputs = FnvHashSet::default();
+        resolve_transaction(transaction, &mut seen_inputs, |x| self.cell(x))
     }
 
-    fn resolve_transaction_at(
+    fn resolve_transaction_at<F: Fn(&OutPoint) -> Option<bool>>(
         &self,
         transaction: &Transaction,
-        parent: &H256,
+        is_spent: F,
     ) -> ResolvedTransaction {
-        let mut seen_inputs = HashSet::new();
-
-        let input_cells = transaction
-            .input_pts()
-            .iter()
-            .map(|input| {
-                if seen_inputs.insert(input.clone()) {
-                    self.cell_at(input, parent)
-                } else {
-                    CellStatus::Dead
-                }
-            })
-            .collect();
-
-        let dep_cells = transaction
-            .dep_pts()
-            .iter()
-            .map(|dep| {
-                if seen_inputs.insert(dep.clone()) {
-                    self.cell_at(dep, parent)
-                } else {
-                    CellStatus::Dead
-                }
-            })
-            .collect();
-
-        ResolvedTransaction {
-            transaction: transaction.clone(),
-            input_cells,
-            dep_cells,
-        }
+        let mut seen_inputs = FnvHashSet::default();
+        resolve_transaction(transaction, &mut seen_inputs, |x| {
+            self.cell_at(x, |o| is_spent(o))
+        })
     }
 
     fn resolve_transaction_unknown_inputs(&self, resolved_transaction: &mut ResolvedTransaction) {
-        for (out_point, state) in resolved_transaction.transaction.out_points_iter().zip(
-            resolved_transaction
-                .dep_cells
-                .iter_mut()
-                .chain(&mut resolved_transaction.input_cells),
-        ) {
-            if *state == CellStatus::Unknown {
-                *state = self.cell(out_point);
+        resolve_transaction_unknown_inputs(resolved_transaction, |x| self.cell(x))
+    }
+}
+
+pub fn resolve_transaction<F: Fn(&OutPoint) -> CellStatus>(
+    transaction: &Transaction,
+    seen_inputs: &mut FnvHashSet<OutPoint>,
```

Comment:
`seen_inputs` is always initialized as an empty set, can we remove it from function params?

Reply:
No, seen_inputs is not always initialized(in block_verify.rs).

Reply:
> No, seen_inputs is not always initialized(eg in block_verify.rs).

It's initialized as an empty set also in block_verify.rs line 395:

```
let mut seen_inputs = FnvHashSet::default();
```

Reply:
but it is passed in `resolve_transaction` many times.